File: src/core/builder.ts
import * as graphlib from '../graphlib';
import * as types from './types';
import { DepGraphImpl } from './dep-graph';
import { validatePackageURL } from './validate-graph';

export { DepGraphBuilder };

class DepGraphBuilder {
  get rootNodeId(): string {
    return this._rootNodeId;
  }

  private static _getPkgId(pkg: types.Pkg): string {
    return `${pkg.name}@${pkg.version || ''}`;
  }

  private _pkgs: { [pkgId: string]: types.PkgInfo } = {};
  private _pkgNodes: { [pkgId: string]: Set<string> } = {};

  private _graph: graphlib.Graph;
  private _pkgManager: types.PkgManager;

  private _rootNodeId: string;
  private _rootPkgId: string;

  public constructor(pkgManager: types.PkgManager, rootPkg?: types.PkgInfo) {
    const graph = new graphlib.Graph({
      directed: true,
      multigraph: false,
      compound: false,
    });
    if (!rootPkg) {
      rootPkg = {
        name: '_root',
        version: '0.0.0',
      };
    }

    this._rootNodeId = 'root-node';
    this._rootPkgId = DepGraphBuilder._getPkgId(rootPkg);
    this._pkgs[this._rootPkgId] = rootPkg;

    graph.setNode(this._rootNodeId, { pkgId: this._rootPkgId });
    this._pkgNodes[this._rootPkgId] = new Set([this._rootNodeId]);

    this._graph = graph;
    this._pkgManager = pkgManager;
  }

  public getPkgs(): types.PkgInfo[] {
    return Object.values(this._pkgs);
  }

  // TODO: this can create disconnected nodes
  public addPkgNode(
    pkgInfo: types.PkgInfo,
    nodeId: string,
    nodeInfo?: types.NodeInfo,
  ) {
    if (nodeId === this._rootNodeId) {
      throw new Error('DepGraphBuilder.addPkgNode() cant override root node');
    }

    validatePackageURL(pkgInfo);

    const pkgId = DepGraphBuilder._getPkgId(pkgInfo);

    this._pkgs[pkgId] = pkgInfo;
    this._pkgNodes[pkgId] = this._pkgNodes[pkgId] || new Set();
    this._pkgNodes[pkgId].add(nodeId);

    this._graph.setNode(nodeId, { pkgId, info: nodeInfo });
    return this;
  }

  // TODO: this can create cycles
  public connectDep(parentNodeId: string, depNodeId: string) {
    if (!this._graph.hasNode(parentNodeId)) {
      throw new Error('parentNodeId does not exist');
    }

    if (!this._graph.hasNode(depNodeId)) {
      throw new Error('depNodeId does not exist');
    }

    this._graph.setEdge(parentNodeId, depNodeId);
    return this;
  }

  public build(): types.DepGraph {
    return new DepGraphImpl(
      this._graph,
      this._rootNodeId,
      this._pkgs,
      this._pkgNodes,
      this._pkgManager,
    );
  }
}

File: go/pkg/depgraph/builder.go
package depgraph

import (
	"errors"
	"fmt"

	"github.com/elliotchance/orderedmap/v3"
)

type Builder struct {
	schemaVersion string
	rootNodeID    string
	rootPkgID     string
	pkgManager    *PkgManager
	pkgs          *orderedmap.OrderedMap[string, *Pkg]
	nodes         *orderedmap.OrderedMap[string, *Node]
}

const (
	schemaVersion = "1.3.0"
	rootNodeID    = "root-node"
)

func NewBuilder(pkgManager *PkgManager, rootPkg *PkgInfo) (*Builder, error) {
	if pkgManager == nil {
		return nil, errors.New("cannot create builder without a package manager")
	}

	if rootPkg == nil {
		rootPkg = &PkgInfo{
			Name:    "_root",
			Version: "unknown",
		}
	}

	b := &Builder{
		schemaVersion: schemaVersion,
		pkgManager:    pkgManager,
		rootNodeID:    rootNodeID,
		rootPkgID:     getPkgID(rootPkg),
		pkgs:          orderedmap.NewOrderedMap[string, *Pkg](),
		nodes:         orderedmap.NewOrderedMap[string, *Node](),
	}

	b.addNode(b.rootNodeID, rootPkg)

	return b, nil
}

func (b *Builder) Build() *DepGraph {
	dg := &DepGraph{
		SchemaVersion: b.schemaVersion,
		PkgManager:    *b.pkgManager,
		Pkgs:          b.GetPkgs(),
		rootPkg:       b.pkgs.GetOrDefault(b.rootPkgID, nil),
		pkgIdx:        make(map[string]*Pkg),
		Graph: Graph{
			RootNodeID: b.rootNodeID,
			Nodes:      make([]Node, b.nodes.Len()),
		},
	}

	var i int
	for _, node := range b.nodes.AllFromFront() {
		nodeID := node.NodeID
		pkgID := node.PkgID
		deps := node.Deps
		info := node.Info
		pkg := b.pkgs.GetOrDefault(pkgID, nil)

		dg.pkgIdx[pkg.ID] = pkg
		dg.Graph.Nodes[i] = Node{
			NodeID: nodeID,
			PkgID:  pkgID,
			Info:   info,
			Deps:   deps,
		}

		i++
	}

	return dg
}

func (b *Builder) GetPkgManager() *PkgManager {
	return b.pkgManager
}

func (b *Builder) GetPkgs() []Pkg {
	pkgs := make([]Pkg, 0, b.pkgs.Len())

	for _, pkg := range b.pkgs.AllFromFront() {
		pkgs = append(pkgs, *pkg)
	}

	return pkgs
}

func (b *Builder) GetRootNode() *Node {
	return b.nodes.GetOrDefault(b.rootNodeID, nil)
}

type nodeOpt = func(*Node)

func WithNodeInfo(info *NodeInfo) nodeOpt {
	return func(node *Node) {
		node.Info = info
	}
}

func (b *Builder) AddNode(nodeID string, pkgInfo *PkgInfo, opts ...nodeOpt) *Node {
	node := b.addNode(nodeID, pkgInfo)

	for _, opt := range opts {
		opt(node)
	}

	return node
}

func (b *Builder) addNode(nodeID string, pkgInfo *PkgInfo) *Node {
	if n, ok := b.nodes.Get(nodeID); ok {
		return n
	}
	pkgID := getPkgID(pkgInfo)

	b.pkgs.Set(pkgID, &Pkg{
		ID:   pkgID,
		Info: *pkgInfo,
	})

	b.nodes.Set(nodeID, &Node{
		NodeID: nodeID,
		PkgID:  pkgID,
		Deps:   make([]Dependency, 0),
	})

	return b.nodes.GetOrDefault(nodeID, nil)
}

func (b *Builder) ConnectNodes(parentNodeID, childNodeID string) error {
	parentNode, ok := b.nodes.Get(parentNodeID)
	if !ok {
		return fmt.Errorf("cound not find parent node %s", parentNodeID)
	}

	childNode, ok := b.nodes.Get(childNodeID)
	if !ok {
		return fmt.Errorf("cound not find child node %s", childNodeID)
	}

	parentNode.Deps = append(parentNode.Deps, Dependency{
		NodeID: childNode.NodeID,
	})

	return nil
}

func getPkgID(pkgInfo *PkgInfo) string {
	return fmt.Sprintf("%s@%s", pkgInfo.Name, pkgInfo.Version)
}
